docs: add a tutorial for malware detection check (#816)
oracle/macaron: docs/source/pages/tutorials/detect_malicious_package.rst

.. Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
.. Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

.. _detect-malicious-package:

----------------------------
Detecting malicious packages
----------------------------

> In this tutorial we show how to use Macaron to find malicious packages. Imagine you have found a Python package and you would like to add this as a dependency to your project, but you are not sure whether you can trust the maintainers or not. You can run Macaron to see if it can detect any malicious behavior. Note that Macaron is an analysis tool and can either miss a malicious behavior or report a false positive. So, while we encourage you to use Macaron, you might still need to triage the results manually.

Reviewer: ```suggestion
In this tutorial we show how to use Macaron to find malicious packages. Imagine you have found a Python package that you would like to add as a dependency to your project, but whose maintainers you are unsure about. In this case, you can run Macaron to see if it can detect any malicious behavior. Note that Macaron is an analysis tool and can either miss malicious behavior or report a false positive. So while we encourage you to use Macaron, you might still need to triage the results manually.
```

I also think the "Note that Macaron..." part would be better as an actual note.

Author: I think the note is fine as it is because this paragraph is at the top and is fairly short.

See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

.. list-table::
   :widths: 25
   :header-rows: 1

   * - Supported packages
   * - Python packages (PyPI)

.. contents:: :local:

**********
Background
**********

> Detecting malicious behavior in open-source software has been a focus for the `Open Source Security Foundation <https://github.com/ossf>`_ (OpenSSF) community in recent years. One significant initiative is :term:`SLSA`, which offers practical recommendations to enhance the integrity of software packages and infrastructure. Macaron is designed to detect poorly maintained or malicious packages by implementing checks inspired by the SLSA specification. However, some forms of attacks currently fall outside the scope of SLSA version 1—notably, SLSA doesn't address the issue of malicious maintainers. Our primary goal is to make it more difficult for malicious actors to compromise critical supply chains and infrastructure. To achieve this, we're developing new methods to detect when maintainers of open source projects are untrustworthy and deliberately spread malware.

Reviewer: ```suggestion
Detecting malicious behavior in open-source software has been a focus for the `Open Source Security Foundation <https://github.com/ossf>`_ (OpenSSF) community in recent years. One significant initiative is :term:`SLSA`, which offers practical recommendations to enhance the integrity of software packages and infrastructure. Macaron is designed to detect poorly maintained or malicious packages by implementing checks inspired by the SLSA specification. However, some forms of attacks currently fall outside the scope of SLSA version 1—notably, SLSA doesn't address the issue of malicious maintainers. Our primary goal is to make it more difficult for malicious actors to compromise critical supply chains and infrastructure. To achieve this, we're developing new methods to detect when maintainers of open source projects are untrustworthy and deliberately spreading malware.
```

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

        * Docker
        * the ``run_macaron.sh``  script to run the Macaron image.

    .. note:: At the moment, Docker alternatives (e.g. podman) are not supported.

    You also need to provide Macaron with a GitHub token through the ``GITHUB_TOKEN``  environment variable.

    To obtain a GitHub Token:

    * Go to ``GitHub settings`` → ``Developer Settings`` (at the bottom of the left side pane) → ``Personal Access Tokens`` → ``Fine-grained personal access tokens`` → ``Generate new token``. Give your token a name and an expiry period.
    * Under ``"Repository access"``, choosing ``"Public Repositories (read-only)"`` should be good enough in most cases.

    Now you should be good to run Macaron. For more details, see the documentation :ref:`here <prepare-github-token>`.

***********
Run Macaron
***********

In this tutorial, we run Macaron on the ``django`` Python package as an example with and without its dependencies to check malicious behavior and apply a policy to fail if the corresponding check fails.

'''''''''''''''''''''''''''''''''''''
Analyzing django without dependencies
'''''''''''''''''''''''''''''''''''''

> First, we need to run the ``analyze`` command of Macaron to run a number of :ref:`checks <checks>` on the ``django`` package. In this tutorial, we are interested in the results of the ``mcn_detect_malicious_metadata_1``. Check :ref:`this tutorial <include_exclude_checks>` if you would like to exclude other checks.

Reviewer 2: ```suggestion
First, we need to run the ``analyze`` command of Macaron to run a number of :ref:`checks <checks>` on the ``django`` package. In this tutorial, we are interested in the results of the ``mcn_detect_malicious_metadata_1`` check. Check :ref:`this tutorial <include_exclude_checks>` if you would like to exclude other checks.
```

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

.. code-block:: shell

  ./run_macaron.sh analyze -purl pkg:pypi/django@5.0.6 --skip-deps

.. note:: By default, Macaron clones the repositories and creates output files under the ``output`` directory. To understand the structure of this directory please see :ref:`Output Files Guide <output_files_guide>`.

.. code-block:: shell

  open output/reports/pypi/django/django.html

.. _fig_django-malware-check:

.. figure:: ../../_static/images/tutorial_django_5.0.6_detect_malicious_metadata_check.png
   :alt: Check ``mcn_detect_malicious_metadata_1`` result for ``django@5.0.6``
   :align: center

> |

Reviewer 2: I wonder if this character is intended here.

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

> The image above shows the result of the ``mcn_detect_malicious_metadata_1`` check for ``django@5.0.6``. The check has passed, which means this package is not malicious. If a package is malicious, this check fails, and if the ecosystem is not supported, the check returns ``UKNOWN``. You can also see the result of individual heuristics applied in this check under the ``Justification`` column.

Reviewer: ```suggestion
The image above shows the result of the ``mcn_detect_malicious_metadata_1`` check for ``django@5.0.6``. The check has passed, which means this package is not malicious. If a package is malicious, this check fails. If the ecosystem is not supported, the check returns ``UNKNOWN``. You can also see the result of individual heuristics applied in this check under the ``Justification`` column.
```

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

> Now we can write a policy to enforce that all versions of ``django`` pass the ``mcn_detect_malicious_metadata_1`` check. The policy will be enforced against the ``analyze`` command run outputs cached in the local database at ``output/macaron.db``.

Reviewer: ```suggestion
Now we can write a policy to ensure that all versions of ``django`` pass the ``mcn_detect_malicious_metadata_1`` check. The policy will be enforced against the output of the ``analyze`` command that is cached in the local database at ``output/macaron.db``.
```

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

.. code-block:: shell

  ./run_macaron.sh verify-policy --database output/macaron.db --file policy.dl

Where the policy looks like below:

.. code-block:: prolog

  #include "prelude.dl"

  Policy("check-django", component_id, "Check django artifacts.") :-
      check_passed(component_id, "mcn_detect_malicious_metadata_1").

  apply_policy_to("check-django", component_id) :-
      is_component(component_id, purl),
      match("pkg:pypi/django@.*", purl).

The ``match`` constraint in this policy allows us to apply the policy on all versions of ``django``. The result of this command should show that the policy succeeds with a zero exit code:

.. code-block:: javascript

    passed_policies
        ['check-django']
    component_satisfies_policy
        ['1', 'pkg:pypi/django@5.0.6', 'check-django']
    failed_policies
    component_violates_policy

> Note that the ``match`` constraint applies a regex and can be expanded to enforce the ``mcn_detect_malicious_metadata_1`` check to pass on all Python packages analyzed so far by Macaron:

Reviewer: ```suggestion
Note that the ``match`` constraint applies a regex pattern and can be expanded to ensure the ``mcn_detect_malicious_metadata_1`` check passes on all Python packages analyzed so far by Macaron:
```

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

.. code-block:: prolog

>   apply_policy_to("check-django", component_id) :-

Reviewer 2: I think in this case, we should rename the policy to `check-all-pypi-packages` or something to prevent confusion:
```
.. code-block:: prolog
  #include "prelude.dl"
  Policy("check-all-pypi-packages", component_id, "Check django artifacts.") :-
      check_passed(component_id, "mcn_detect_malicious_metadata_1").
  apply_policy_to("check-all-pypi-packages", component_id) :-
      is_component(component_id, purl),
      match("pkg:pypi.*", purl).
```

Author: This modification is suggested as a variation of the previous policy and the policy name should not change.

      is_component(component_id, purl),
      match("pkg:pypi.*", purl).

Note that if a policy fails to pass, Macaron returns a none-zero error code.

''''''''''''''''''''''''''''''''''
Analyzing django with dependencies
''''''''''''''''''''''''''''''''''

> Macaron supports analyzing dependencies of a package and runs the same set of checks on the dependencies as the main target package. To analyze the dependencies of ``django@5.0.6`` Python package, you can either :ref:`generate an SBOM <python-sbom>` yourself or :ref:`point Macaron to a virtual environment <python-venv-deps>` where ``django`` is installed.

Reviewer: ```suggestion
Macaron supports analyzing dependencies of a package, and will run the same set of checks on the dependencies as for the main target package. To analyze the dependencies of the ``django@5.0.6`` Python package, you can either :ref:`generate an SBOM <python-sbom>` yourself or :ref:`point Macaron to a virtual environment <python-venv-deps>` where ``django`` is installed.
```

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

Let's assume ``/tmp/.django_venv`` is the virtual environment where ``django@5.0.6`` is installed. Run Macaron as follows to analyze ``django`` and its dependencies.

.. code-block:: shell

  ./run_macaron.sh analyze -purl pkg:pypi/django@5.0.6 --python-venv "/tmp/.django_venv"

By default Macaron only checks the direct dependencies. To turn on recursive dependency analysis, add the following to the ``configurations.ini`` file:

.. code-block:: ini

  [dependency.resolver]
  recursive = True

And pass that to the ``analyze`` command:

.. code-block:: shell

  ./run_macaron.sh --defaults-path configurations.ini analyze -purl pkg:pypi/django@5.0.6 --python-venv "/tmp/.django_venv"

To learn more about changing configurations see :ref:`here <change-config>`.

> Now we can enforce the policy below to ensure that the ``mcn_detect_malicious_metadata_1`` check always passes on ``django`` and its dependencies and none of the dependencies have malicious behavior.

Reviewer 2: ```suggestion
Now we can enforce the policy below to ensure that the ``mcn_detect_malicious_metadata_1`` check always passes on ``django`` and its dependencies, indicating none of the dependencies have malicious behavior.
```
I think using "and" create a confusion that it's a third thing that the policy ensures, whereas it's more of a conclusion from the policy result (as least for me).

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

.. code-block:: prolog

  #include "prelude.dl"

  Policy("check-dependencies", component_id, "Check the dependencies of django.") :-

>     is_component(component_id, _),

Reviewer 2: Why does we need `is_component(component_id, _)` while it's not needed in this policy file https://github.com/oracle/macaron/blob/24090493071ca4d9e3e5998338935f6410e9f41c/tests/integration/cases/django_with_dep_resolution_virtual_env_as_input/policy-recursive-deps.dl#L6-L9.
Or are they supposed to be different from each other ? :thinking:

Author: Good catch. I had forgotten to update the policy in docs after making changes in the test. See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

    transitive_dependency(component_id, dependency),

>     check_passed(dependency, "mcn_detect_malicious_metadata_1").

Reviewer 2: I think we should have another rule for `component_id` passing `mcn_detect_malicious_metadata_1`.

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

  apply_policy_to("check-dependencies", component_id) :-
      is_component(component_id, purl),

>       match("pkg:pypi.*", purl).

Reviewer 2: I think we should match the PURL to `django` instead. If we use `pkg:pypi,*` this policy will be run against all PYPI components which include the direct dependencies. 
If a component doesn't have any dependencies, it will fail the policy, which I don't think is desirable. 
For example, this could be the result of using this policy:
```
component_satisfies_policy
    ['1', 'pkg:pypi/django@5.0.6', 'check-dependencies']
passed_policies
component_violates_policy
    ['2', 'pkg:pypi/asgiref@3.8.1', 'check-dependencies']
    ['3', 'pkg:pypi/sqlparse@0.5.1', 'check-dependencies']
    ['4', 'pkg:pypi/pip@22.3.1', 'check-dependencies']
failed_policies
    ['check-dependencies']
```

Author: Yes, this was meant to be matching `django` as you can see in the related integration policy `[policy-recursive-deps.dl](https://github.com/oracle/macaron/blob/24090493071ca4d9e3e5998338935f6410e9f41c/tests/integration/cases/django_with_dep_resolution_virtual_env_as_input/policy-recursive-deps.dl#L6-L9)`.

See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

> As you can see below, the policy passes for ``django`` and all its transitive dependencies.

Reviewer 2: ```suggestion
As you can see below, the policy passes for ``django`` because it doesn't detect malicious behavior for `django` and all its transitive dependencies.
```
Because the target of the Policy is `django`, using `and all its transitive dependencies.` create an impression for me that each transitive dependency is also a target.

Reviewer: I agree, except it should end with `or any of its transitive dependencies.`

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

.. code-block:: javascript

  passed_policies
      ['check-dependencies']
  component_satisfies_policy
      ['1', 'pkg:pypi/django@5.0.6', 'check-dependencies']
  failed_policies
  component_violates_policy

***********
Future Work
***********

> We are actively working on the malware detection analysis in Macaron to improve the precision and support more ecosystems. An upcoming feature is a more advanced source code analysis. Stay tuned and feel free to contribute to improve this check.

Reviewer: ```suggestion
We are actively working on the malware detection analysis check in Macaron — to improve precision, support more ecosystems, and in particular, perform more advanced source code analysis. Stay tuned and feel free to contribute to improve this check.
```

Author: See commit [728d753](https://github.com/oracle/macaron/pull/816/commits/728d753678ee948b303d9b2fb9c672f245ed7ee0)

      is_component(component_id, purl),
      match("pkg:pypi.*", purl).

+++++++++++++++++++++++++++++++++++++++
Verification Summary Attestation report
+++++++++++++++++++++++++++++++++++++++

> Additionally, Macaron generates a Verification Summary Attestation (:term:`VSA`) report that contains the policy and information about the analyzed artifact. See :ref:`this page <vsa>` for more details. For instance, the VSA report for the ``check-django`` policy shown above can be viewed by running the command below:

Reviewer: ```suggestion
Additionally, Macaron generates a Verification Summary Attestation (:term:`VSA`) report that contains the policy, and information about the analyzed artifact. See :ref:`this page <vsa>` for more details. For instance, the VSA report for the ``check-django`` policy shown above can be viewed by running this command:
```

Author: See commit [003f728](https://github.com/oracle/macaron/pull/816/commits/003f7289e091ae385a59712b6711c3edd2f39c91)

  failed_policies
  component_violates_policy

''''''''''''''''''''''''''''''''''''''''
Require a confidence level in the policy
''''''''''''''''''''''''''''''''''''''''

> Macaron also provides a confidence score for each check result, represented as a value between ``0`` and ``1`` inclusive, respectively. You can incorporate this score into your policy to ensure checks meet a required level of confidence. Currently, Macaron :class:`has these confidence levels <macaron.slsa_analyzer.checks.check_result.Confidence>`. For instance, you might adjust the :ref:`check-dependencies policy shown earlier <django_with_deps>` to require that the ``mcn_detect_malicious_metadata_1`` check passes with a high confidence, i.e., ``1``:

Reviewer: ```suggestion
Macaron also provides a confidence score for each check result, represented as a value ranging from ``0`` to ``1`` (inclusively). You can incorporate this score into your policy to ensure checks meet a required level of confidence. Currently, Macaron :class:`has these confidence levels <macaron.slsa_analyzer.checks.check_result.Confidence>`. For instance, you might adjust the :ref:`check-dependencies policy shown earlier <django_with_deps>` to require that the ``mcn_detect_malicious_metadata_1`` check passes with a high confidence, i.e., ``1``:
```

Author: See commit [003f728](https://github.com/oracle/macaron/pull/816/commits/003f7289e091ae385a59712b6711c3edd2f39c91)
